snowflake/src/test/java/org/apache/iceberg/snowflake/JdbcSnowflakeClientTest.java

```diff
@@ -367,15 +535,66 @@ public void testListIcebergTablesInSchema() throws SQLException {
 
   /**
    * Any unexpected SQLException from the underlying connection will propagate out as an
-   * UncheckedSQLException when listing tables.
+   * UncheckedSQLException when listing tables at Root level
    */
   @Test
-  public void testListIcebergTablesSQLException() throws SQLException, InterruptedException {
-    when(mockClientPool.run(any(ClientPool.Action.class)))
-        .thenThrow(new SQLException("Fake SQL exception"));
-    Assertions.assertThatExceptionOfType(UncheckedSQLException.class)
-        .isThrownBy(() -> snowflakeClient.listIcebergTables(SnowflakeIdentifier.ofDatabase("DB_1")))
-        .withStackTraceContaining("Fake SQL exception");
+  public void testListIcebergTablesSQLExceptionAtRootLevel()
+      throws SQLException, InterruptedException {
+    generateExceptionAndAssert(
+        () -> snowflakeClient.listIcebergTables(SnowflakeIdentifier.ofRoot()),
+        new SQLException(String.format("SQL exception with Error Code %d", 0), "2000", 0, null),
+        UncheckedSQLException.class);
+  }
+
+  /**
+   * Any unexpected SQLException with specific error codes from the underlying connection will
+   * propagate out as a NoSuchNamespaceException when listing tables at Database level
+   */
+  @Test
+  public void testListIcebergTablesSQLExceptionAtDatabaseLevel()
+      throws SQLException, InterruptedException {
+    for (Integer errorCode : DATABASE_NOT_FOUND_ERROR_CODES) {
+      generateExceptionAndAssert(
+          () -> snowflakeClient.listIcebergTables(SnowflakeIdentifier.ofDatabase("DB_1")),
+          new SQLException(
+              String.format("SQL exception with Error Code %d", errorCode),
+              "2000",
+              errorCode,
+              null),
+          NoSuchNamespaceException.class);
+    }
+  }
+
+  /**
+   * Any unexpected SQLException with specific error codes from the underlying connection will
+   * propagate out as a NoSuchNamespaceException when listing tables at Schema level
+   */
+  @Test
+  public void testListIcebergTablesSQLExceptionAtSchemaLevel()
+      throws SQLException, InterruptedException {
+    for (Integer errorCode : SCHEMA_NOT_FOUND_ERROR_CODES) {
```

Comment:
TBH using `generateExceptionAndAssert()` makes the code quite difficult to read and reason about. I understand the argument for reusing the same code, but I think tests are clearer to read when we're asserting everything in the test method, similar to below: 

```
 @Test
  public void testListIcebergTablesSQLExceptionAtSchemaLevel()
      throws SQLException, InterruptedException {
    for (Integer errorCode : SCHEMA_NOT_FOUND_ERROR_CODES) {
      SQLException injectedException =
          new SQLException(
              String.format("SQL exception with Error Code %d", errorCode),
              "2000",
              errorCode,
              null);
      when(mockClientPool.run(any(ClientPool.Action.class))).thenThrow(injectedException);

      Assertions.assertThatExceptionOfType(NoSuchNamespaceException.class)
          .isThrownBy(
              () ->
                  snowflakeClient.listIcebergTables(
                      SnowflakeIdentifier.ofSchema("DB_1", "SCHEMA_1")))
          .withMessageContaining("Identifier not found: 'SCHEMA: 'DB_1.SCHEMA_1''")
          .withCause(injectedException);
    }
  }
```

Note that we're using `.withMessageContaining` to make sure we're getting the right error message, and `.withCause` allows to make sure we get the correct exception in the cause.